velox/substrait/SubstraitToVeloxPlan.cpp

```diff
@@ -282,49 +342,115 @@ std::shared_ptr<const core::PlanNode> SubstraitVeloxPlanConverter::toVeloxPlan(
   }
   auto outputType = ROW(std::move(outNames), std::move(veloxTypeList));
 
-  auto tableScanNode = std::make_shared<core::TableScanNode>(
-      nextPlanNodeId(), outputType, tableHandle, assignments);
-  return tableScanNode;
+  if (sRead.has_virtual_table()) {
+    return toVeloxPlan(sRead, pool, outputType);
+
+  } else {
+    auto tableScanNode = std::make_shared<core::TableScanNode>(
+        nextPlanNodeId(), outputType, tableHandle, assignments);
+    return tableScanNode;
+  }
+}
+
+std::shared_ptr<const core::PlanNode> SubstraitVeloxPlanConverter::toVeloxPlan(
+    const ::substrait::ReadRel& readRel,
+    facebook::velox::memory::MemoryPool* pool,
+    std::shared_ptr<const RowType> type) {
+  ::substrait::ReadRel_VirtualTable readVirtualTable = readRel.virtual_table();
+  int64_t numRows = readVirtualTable.values_size();
+  int64_t numColumns = type->size();
+  int64_t valueFieldNums = readVirtualTable.values(numRows - 1).fields_size();
+  std::vector<RowVectorPtr> vectors;
+  vectors.reserve(numRows);
+
+  int64_t batchSize = valueFieldNums / numColumns;
+
+  for (int64_t row = 0; row < numRows; ++row) {
+    std::vector<VectorPtr> children;
+    ::substrait::Expression_Literal_Struct rowValue =
+        readRel.virtual_table().values(row);
+    auto fieldSize = rowValue.fields_size();
+    VELOX_CHECK_EQ(fieldSize, batchSize * numColumns);
+
+    const auto& childrenSize = type->children().size();
```

Comment:
This variable is not needed as it is the same as numColumns. Let's remove.

Reply:
sure.